### PR TITLE
Fix: allow empty loop body in no-extra-semi (fixes #3075)

### LINKED ISSUE #3075: `no-extra-semi` advice is ill on loops without a body
```js
for (var i = 0; i <= 5; console.log(i++));
```

[Real life example](https://github.com/nodejs/io.js/blob/c7d8b091623b5dd30977d29f946051c0bf328ad1/lib/cluster.js#L157)

While I agree this is hardly readable and should be refactored, `extra-semi` does incorrectly advice to remove the semicolon here, which would lead to the next line being considered the body of the loop, changing behaviour. The same issue can happen on a `while` loop:
```js
var i;
while (i++ <= 5);
```
[Real life example](https://github.com/nodejs/io.js/blob/c7d8b091623b5dd30977d29f946051c0bf328ad1/lib/url.js#L502)

A `do while` is unaffected because the curly braces are required there. This is on v24.1.

### PR DESCRIPTION
**Pull request by lo1tuma**

### CODE REVIEW

**Comment by btmills** [comment]
LGTM, but waiting to merge until the question about `curly` is resolved in #3075.

File: lib/rules/no-extra-semi.js [hunk 1]
```diff
@@ -40,11 +40,18 @@ module.exports = function(context) {
...
         /**
-         * Reports this empty statement.
+         * Reports this empty statement, except if the parent node is a loop.
          * @param {Node} node - A EmptyStatement node to be reported.
          * @returns {void}
          */
-        "EmptyStatement": report,
+        "EmptyStatement": function(node) {
+            var parent = node.parent,
+                allowedParentTypes = ["ForStatement", "WhileStatement", "DoWhileStatement"];
```
Comment (michaelficarra) [inline lib/rules/no-extra-semi.js]: for-in, for-of

**Comment by ilyavolodin** [comment]
Issue with curly on the same construct: #3077

File: lib/rules/no-extra-semi.js (hunk 1 above)
Comment (lo1tuma) [inline lib/rules/no-extra-semi.js]: done

**Comment by nzakas** [comment]
Lgtm

File: lib/rules/no-extra-semi.js [hunk 2]
```diff
@@ -40,11 +40,18 @@ module.exports = function(context) {
...
         /**
-         * Reports this empty statement.
+         * Reports this empty statement, except if the parent node is a loop.
          * @param {Node} node - A EmptyStatement node to be reported.
          * @returns {void}
          */
-        "EmptyStatement": report,
+        "EmptyStatement": function(node) {
+            var parent = node.parent,
+                allowedParentTypes = ["ForStatement", "ForInStatement", "ForOfStatement", "WhileStatement", "DoWhileStatement"];
```
Comment (michaelficarra) [inline lib/rules/no-extra-semi.js]: WithStatement

**Pull request merged**